File: memcontainers/conformance/test/control_vectors.test.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import {
  decodeDirEntries,
  decodeExecOutcome,
  decodeExecRequest,
  decodeFileStat,
  decodeRelayEvent,
  decodeSvcRequest,
  decodeSvcResponse,
  encodeDirEntries,
  encodeExecOutcome,
  encodeExecRequest,
  encodeFileStat,
  encodeRelayEvent,
  encodeSvcRequest,
  encodeSvcResponse,
  WireError,
} from "../../contracts/gen/ctl.gen.js";

type VectorDoc = {
  schema: number;
  positive: Array<{ message: string; description: string; hex: string }>;
  negative: Array<{ message: string; name: string; hex: string; error: string }>;
};

const te = new TextEncoder();

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) throw new Error(message);
}

function runfile(rel: string | undefined, envVar: string): string {
  if (!rel) throw new Error(`${envVar} is not set`);
  const rf = process.env.RUNFILES_DIR;
  if (!rf) throw new Error("RUNFILES_DIR is not set");
  return join(rf, rel);
}

function vectors(): VectorDoc {
  return JSON.parse(readFileSync(runfile(process.env.MC_CONTROL_VECTORS, "MC_CONTROL_VECTORS"), "utf8"));
}

function hex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

function bytes(hexString: string): Uint8Array {
  assert(hexString.length % 2 === 0, "hex vector must have an even length");
  const out = new Uint8Array(hexString.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = Number.parseInt(hexString.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

function positive(message: string): Uint8Array {
  const found = vectors().positive.find((v) => v.message === message);
  if (!found) throw new Error(`missing positive vector for ${message}`);
  return bytes(found.hex);
}

function assertEqualBytes(actual: Uint8Array, expected: Uint8Array, message: string): void {
  assert(actual.length === expected.length, `${message}: length ${actual.length} !== ${expected.length}`);
  for (let i = 0; i < actual.length; i++) {
    assert(actual[i] === expected[i], `${message}: byte ${i} ${actual[i]} !== ${expected[i]}`);
  }
}

function assertOptionalBytes(actual: Uint8Array | null | undefined, expected: Uint8Array, message: string): void {
  assert(actual !== undefined && actual !== null, `${message}: missing bytes`);
  assertEqualBytes(actual, expected, message);
}

function positiveVectors(): void {
  const exec = {
    cmd: "printf $ALPHA && cat",
    cwd: "/work",
    env: { ZED: "last", ALPHA: "first" },
    stdin: new Uint8Array([112, 97, 121, 108, 111, 97, 100, 10, 0]),
  };
  assert(hex(encodeExecRequest(exec)) === hex(positive("ExecRequest")), "ExecRequest fixture drifted");
  const decodedExec = decodeExecRequest(positive("ExecRequest"));
  assert(decodedExec.cmd === exec.cmd && decodedExec.cwd === exec.cwd, "ExecRequest decoded scalar fields changed");
  assert(decodedExec.env.ALPHA === "first" && decodedExec.env.ZED === "last", "ExecRequest decoded env changed");
  assertOptionalBytes(decodedExec.stdin, exec.stdin, "ExecRequest decoded stdin changed");

  const outcome = { exit_code: 7, stdout: te.encode("out\n"), stderr: te.encode("err\n") };
  assert(hex(encodeExecOutcome(outcome)) === hex(positive("ExecOutcome")), "ExecOutcome fixture drifted");
  const decodedOutcome = decodeExecOutcome(positive("ExecOutcome"));
  assert(decodedOutcome.exit_code === 7, "ExecOutcome exit code changed");
  assertEqualBytes(decodedOutcome.stdout, outcome.stdout, "ExecOutcome stdout changed");
  assertEqualBytes(decodedOutcome.stderr, outcome.stderr, "ExecOutcome stderr changed");

  const stat = { size: 12345, is_dir: false, is_symlink: true, nlink: 2, mode: 0o120777 };
  assert(hex(encodeFileStat(stat)) === hex(positive("FileStat")), "FileStat fixture drifted");
  const decodedStat = decodeFileStat(positive("FileStat"));
  assert(decodedStat.size === 12345 && decodedStat.is_symlink && decodedStat.nlink === 2, "FileStat decoded fields changed");

  const entries = {
    entries: [
      { name: "a.txt", is_dir: false, is_symlink: false },
      { name: "link", is_dir: false, is_symlink: true },
      { name: "sub", is_dir: true, is_symlink: false },
    ],
  };
  assert(hex(encodeDirEntries(entries)) === hex(positive("DirEntries")), "DirEntries fixture drifted");
  const decodedEntries = decodeDirEntries(positive("DirEntries"));
  assert(decodedEntries.entries.length === 3, "DirEntries decoded length changed");
  assert(decodedEntries.entries[1]?.name === "link" && decodedEntries.entries[1]?.is_symlink, "DirEntries symlink changed");

  const svcRequest = { service: "kv", request: te.encode("put\0answer\0forty-two") };
  assert(hex(encodeSvcRequest(svcRequest)) === hex(positive("SvcRequest")), "SvcRequest fixture drifted");
  const decodedSvcRequest = decodeSvcRequest(positive("SvcRequest"));
  assert(decodedSvcRequest.service === "kv", "SvcRequest service changed");
  assertEqualBytes(decodedSvcRequest.request, svcRequest.request, "SvcRequest body changed");

  const svcResponse = { status: 0, body: te.encode("42") };
  assert(hex(encodeSvcResponse(svcResponse)) === hex(positive("SvcResponse")), "SvcResponse fixture drifted");
  const decodedSvcResponse = decodeSvcResponse(positive("SvcResponse"));
  assert(decodedSvcResponse.status === 0, "SvcResponse status changed");
  assertEqualBytes(decodedSvcResponse.body, svcResponse.body, "SvcResponse body changed");

  const relay = {
    kind: "host_call",
    handle: 42,
    name: "tool.exec",
    body: new Uint8Array([0, 1, 2, 255]),
    args_digest: "sha256:0123456789abcdef",
  };
  assert(hex(encodeRelayEvent(relay)) === hex(positive("RelayEvent")), "RelayEvent fixture drifted");
  const decodedRelay = decodeRelayEvent(positive("RelayEvent"));
  assert(decodedRelay.kind === "host_call" && decodedRelay.handle === 42, "RelayEvent scalar fields changed");
  assert(decodedRelay.name === "tool.exec", "RelayEvent name changed");
  assertOptionalBytes(decodedRelay.body, relay.body, "RelayEvent body changed");
  assert(decodedRelay.args_digest === "sha256:0123456789abcdef", "RelayEvent args digest changed");
}

function negativeVectors(): void {
  const expectedMessages: Record<string, string> = {
    WrongMessage: "wrong message id",
    UnsupportedVersion: "unsupported message version",
    NonCanonicalMap: "non-canonical strmap",
    TrailingBytes: "trailing bytes",
  };
  for (const vector of vectors().negative) {
    try {
      if (vector.message === "ExecRequest") {
        decodeExecRequest(bytes(vector.hex));
      } else {
        throw new Error(`unhandled negative vector message ${vector.message}`);
      }
    } catch (error) {
      assert(error instanceof WireError, `${vector.message}:${vector.name}: expected WireError`);
      const expected = expectedMessages[vector.error];
      assert(expected !== undefined, `${vector.message}:${vector.name}: unknown expected error ${vector.error}`);
      assert(
        error.message === expected,
        `${vector.message}:${vector.name}: ${JSON.stringify(error.message)} !== ${JSON.stringify(expected)}`,
      );
      continue;
    }
    throw new Error(`${vector.message}:${vector.name}: decoded unexpectedly`);
  }
}

positiveVectors();
negativeVectors();
console.log("control conformance vectors match Rust/TS generated codecs");

File: memcontainers/conformance/src/control_vectors_test.rs
use std::collections::BTreeMap;

use ctl_rust::{
    DirEntries, DirEntry, ExecOutcome, ExecRequest, FileStat, RelayEvent, SvcRequest, SvcResponse,
    WireError,
};
use serde_json::Value;

fn runfile(path: &str) -> Vec<u8> {
    let r = runfiles::Runfiles::create().expect("runfiles unavailable");
    let p = r
        .rlocation(path)
        .unwrap_or_else(|| panic!("{path} not found in runfiles"));
    std::fs::read(&p).unwrap_or_else(|e| panic!("reading {}: {e}", p.display()))
}

fn vectors() -> Value {
    serde_json::from_slice(&runfile(
        "_main/memcontainers/conformance/control_vectors.json",
    ))
    .expect("parse control_vectors.json")
}

fn hex_to_bytes(hex: &str) -> Vec<u8> {
    assert_eq!(hex.len() % 2, 0, "hex vector must have an even length");
    (0..hex.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&hex[i..i + 2], 16).expect("hex byte"))
        .collect()
}

fn positive(message: &str) -> Vec<u8> {
    vectors()["positive"]
        .as_array()
        .expect("positive vectors")
        .iter()
        .find(|case| case["message"].as_str() == Some(message))
        .map(|case| hex_to_bytes(case["hex"].as_str().expect("hex")))
        .unwrap_or_else(|| panic!("missing positive vector for {message}"))
}

fn map(pairs: &[(&str, &str)]) -> BTreeMap<String, String> {
    pairs
        .iter()
        .map(|(k, v)| ((*k).to_string(), (*v).to_string()))
        .collect()
}

#[test]
fn rust_control_codecs_match_the_shared_positive_vectors() {
    let exec = ExecRequest {
        cmd: "printf $ALPHA && cat".to_string(),
        cwd: Some("/work".to_string()),
        env: map(&[("ZED", "last"), ("ALPHA", "first")]),
        stdin: Some(b"payload\n\0".to_vec()),
    };
    assert_eq!(exec.encode(), positive("ExecRequest"));
    let decoded = ExecRequest::decode(&positive("ExecRequest")).unwrap();
    assert_eq!(decoded.cmd, exec.cmd);
    assert_eq!(decoded.cwd, exec.cwd);
    assert_eq!(decoded.env, map(&[("ALPHA", "first"), ("ZED", "last")]));
    assert_eq!(decoded.stdin.as_deref(), Some(b"payload\n\0".as_slice()));

    let outcome = ExecOutcome {
        exit_code: 7,
        stdout: b"out\n".to_vec(),
        stderr: b"err\n".to_vec(),
    };
    assert_eq!(outcome.encode(), positive("ExecOutcome"));
    assert_eq!(
        ExecOutcome::decode(&positive("ExecOutcome")).unwrap(),
        outcome
    );

    let stat = FileStat {
        size: 12345,
        is_dir: false,
        is_symlink: true,
        nlink: 2,
        mode: 0o120777,
    };
    assert_eq!(stat.encode(), positive("FileStat"));
    assert_eq!(FileStat::decode(&positive("FileStat")).unwrap(), stat);

    let entries = DirEntries {
        entries: vec![
            DirEntry {
                name: "a.txt".to_string(),
                is_dir: false,
                is_symlink: false,
            },
            DirEntry {
                name: "link".to_string(),
                is_dir: false,
                is_symlink: true,
            },
            DirEntry {
                name: "sub".to_string(),
                is_dir: true,
                is_symlink: false,
            },
        ],
    };
    assert_eq!(entries.encode(), positive("DirEntries"));
    assert_eq!(
        DirEntries::decode(&positive("DirEntries")).unwrap(),
        entries
    );

    let svc_request = SvcRequest {
        service: "kv".to_string(),
        request: b"put\0answer\0forty-two".to_vec(),
    };
    assert_eq!(svc_request.encode(), positive("SvcRequest"));
    assert_eq!(
        SvcRequest::decode(&positive("SvcRequest")).unwrap(),
        svc_request
    );

    let svc_response = SvcResponse {
        status: 0,
        body: b"42".to_vec(),
    };
    assert_eq!(svc_response.encode(), positive("SvcResponse"));
    assert_eq!(
        SvcResponse::decode(&positive("SvcResponse")).unwrap(),
        svc_response
    );

    let relay = RelayEvent {
        kind: "host_call".to_string(),
        handle: 42,
        name: Some("tool.exec".to_string()),
        body: Some(vec![0, 1, 2, 255]),
        args_digest: Some("sha256:0123456789abcdef".to_string()),
        ..Default::default()
    };
    assert_eq!(relay.encode(), positive("RelayEvent"));
    assert_eq!(RelayEvent::decode(&positive("RelayEvent")).unwrap(), relay);
}

#[test]
fn rust_control_codecs_reject_shared_negative_vectors() {
    for case in vectors()["negative"].as_array().expect("negative vectors") {
        let message = case["message"].as_str().expect("message");
        let name = case["name"].as_str().expect("name");
        let bytes = hex_to_bytes(case["hex"].as_str().expect("hex"));
        let want = case["error"].as_str().expect("error");
        let got = match message {
            "ExecRequest" => ExecRequest::decode(&bytes).expect_err("negative vector decoded"),
            other => panic!("unhandled negative vector message {other}"),
        };
        let got_name = match got {
            WireError::WrongMessage => "WrongMessage",
            WireError::UnsupportedVersion => "UnsupportedVersion",
            WireError::Truncated => "Truncated",
            WireError::TrailingBytes => "TrailingBytes",
            WireError::NonCanonicalMap => "NonCanonicalMap",
            WireError::InvalidUtf8 => "InvalidUtf8",
            WireError::InvalidPresence => "InvalidPresence",
        };
        assert_eq!(got_name, want, "{message}:{name}");
    }
}
